letsencrypt/models/letsencrypt.py

```diff
@@ -2,167 +2,464 @@
+
+        _logger.info(
+            "Certificate expires on %s, which is in %d days, no action needed",
+            expiry,
+            remaining.days,
+        )
+        return False
 
     @api.model
-    def generate_domain_key(self, domain):
-        domain_key = os.path.join(get_data_dir(), '%s.key' % domain)
-        if not os.path.isfile(domain_key):
-            _logger.info('generating rsa domain key for %s', domain)
-            self.call_cmdline([
-                'openssl', 'genrsa', '-out', domain_key,
-                str(DEFAULT_KEY_LENGTH),
-            ])
-        return domain_key
+    def _cron(self):
+        ir_config_parameter = self.env['ir.config_parameter']
+        base_url = ir_config_parameter.get_param('web.base.url', 'localhost')
+        domain = urlparse.urlparse(base_url).hostname
+        cert_file = os.path.join(_get_data_dir(), '%s.crt' % domain)
+
+        domains = self._cascade_domains([domain] + self._get_altnames())
+        for dom in domains:
+            self._validate_domain(dom)
+
+        if not self._should_run(cert_file, domains):
+            return
+
+        account_key = josepy.JWKRSA.load(self._get_key('account.key'))
+        domain_key = self._get_key('%s.key' % domain)
+
+        client = self._create_client(account_key)
+        new_reg = acme.messages.NewRegistration(
+            key=account_key.public_key(), terms_of_service_agreed=True
+        )
+        try:
+            client.new_account(new_reg)
+            _logger.info("Successfully registered.")
+        except acme.errors.ConflictError as err:
+            reg = acme.messages.Registration(key=account_key.public_key())
+            reg_res = acme.messages.RegistrationResource(
+                body=reg, uri=err.location
+            )
+            client.query_registration(reg_res)
+            _logger.info("Reusing existing account.")
+
+        _logger.info('Making CSR for the following domains: %s', domains)
+        csr = acme.crypto_util.make_csr(
+            private_key_pem=domain_key, domains=domains
+        )
+        authzr = client.new_order(csr)
+
+        # For each requested domain name we receive a list of challenges.
+        # We only have to do one from each list.
+        # HTTP challenges are the easiest, so do one of those if possible.
+        # We can do DNS challenges too. There are other types that we don't
+        # support.
+        pending_responses = []
+
+        for authorizations in authzr.authorizations:
+            http_challenges = [
+                challenge
+                for challenge in authorizations.body.challenges
+                if challenge.chall.typ == TYPE_CHALLENGE_HTTP
+            ]
+            other_challenges = [
+                challenge
+                for challenge in authorizations.body.challenges
+                if challenge.chall.typ != TYPE_CHALLENGE_HTTP
+            ]
+            for challenge in http_challenges + other_challenges:
```

Comment:
Hm, doing that automatically would break other use cases, like when not all of your domains go through that DNS provider. DNS challenges are also slower and less reliable.

The rigorous solution could be to add a fully-fledged `letsencrypt.domain` model for per-domain configuration records. That could also make it possible to use different DNS providers for different domains, which you currently can't do. But that would be overengineering.

A "prefer DNS challenges" checkbox seems reasonable enough. I'll add that.

I do wonder whether a self-signed certificate would make more sense if Odoo isn't externally accessible.

Reply:
the setting is VPN + clients that aren't under our control, so self signed won't work properly there. Thanks for the flag, I indeed think a config per domain would be overkill

Reply:
I've implemented this.